File: First year/Second Semester/Graphs/A1/AS1 - python/directed_graph.py
from graph_exceptions import GraphException
import copy
from random import randint


class DirectedGraph:

    def __init__(self, number_of_vertices=0):
        self.number_of_vertices = number_of_vertices
        self.number_of_edges = 0
        self.inbound_edges = {}
        self.outbound_edges = {}
        self.costs = {}

    def __str__(self):
        string = f"DirectedGraph: {self.number_of_vertices} vertices, {self.number_of_edges} edges"
        for vertex, inbound_edges in self.inbound_edges.items():
            for inbound_edge in inbound_edges:
                if not inbound_edges and not self.outbound_edges[vertex]:
                    string += f"\n {vertex} -> isolated vertex"
                    continue

                string += f"\n {inbound_edge} -> {vertex} ->: {self.costs[(inbound_edge, vertex)]}"

        return string

    def get_number_of_vertices(self):
        return self.number_of_vertices

    def set_number_of_vertices(self, number_of_vertices):
        self.number_of_vertices = number_of_vertices

    def get_cost_of_an_edge(self, source, destination):
        if self.check_if_edge_exists(source, destination):
            return self.costs[(source, destination)]
        else:
            raise GraphException("There is no edge from {source} to {destination}")

    def set_cost_of_an_edge(self, source, destination, cost):
        if self.check_if_edge_exists(source, destination):
            self.costs[(source, destination)] = cost
        else:
            self.add_edge(source, destination, cost)

    def get_vertices(self):
        return list(self.inbound_edges.keys())

    def check_if_edge_exists(self, source, destination) -> bool:
        return (source, destination) in self.costs

    def add_vertex(self, vertex):
        if vertex in self.inbound_edges:
            raise GraphException("The vertex already exists")

        self.inbound_edges[vertex] = []
        self.outbound_edges[vertex] = []
        self.number_of_vertices += 1

    def remove_vertex(self, vertex):
        if vertex not in self.inbound_edges:
            raise GraphException("The vertex does not exist")

        if vertex in self.inbound_edges:
            self.number_of_vertices -= 1
            self.number_of_edges -= len(self.inbound_edges[vertex]) + len(self.outbound_edges[vertex])
            for inbound_vertex in self.inbound_edges[vertex]:
                del self.costs[(vertex, inbound_vertex)]
                del self.outbound_edges[inbound_vertex][vertex]
            for outbound_vertex in self.outbound_edges[vertex]:
                del self.costs[(vertex, outbound_vertex)]
                del self.inbound_edges[outbound_vertex][vertex]
            del self.inbound_edges[vertex]
            del self.outbound_edges[vertex]

    def add_edge(self, source, destination, cost):
        if self.check_if_edge_exists(source, destination):
            raise GraphException("The edge already exists")
        try:
            self.add_vertex(source)
        except GraphException:
            pass
        try:
            self.add_vertex(destination)
        except GraphException:
            pass

        if source not in self.outbound_edges:
            self.outbound_edges[source] = []
        if destination not in self.inbound_edges:
            self.inbound_edges[destination] = []

        self.outbound_edges[source].append(destination)
        self.inbound_edges[destination].append(source)
        self.costs[(source, destination)] = cost
        self.number_of_edges += 1

    def remove_edge(self, source, destination):
        if self.check_if_edge_exists(source, destination):
            del self.costs[(source, destination)]
            self.outbound_edges[source].remove(destination)
            self.inbound_edges[destination].remove(source)
            self.number_of_edges -= 1
            return
        raise GraphException("The edge does not exist")

    def get_copy_of_graph(self):
        graph_copy = DirectedGraph(self.number_of_vertices)
        graph_copy.inbound_edges = copy.deepcopy(self.inbound_edges)
        graph_copy.outbound_edges = copy.deepcopy(self.outbound_edges)
        graph_copy.costs = copy.deepcopy(self.costs)
        return graph_copy

    def create_random_graph(self, number_of_vertices, number_of_edges):
        if number_of_edges > number_of_vertices ** 2:
            raise GraphException("Invalid input! The number of edges must be less than the number of vertices squared.")
        if self.number_of_vertices != 0:
            raise GraphException("The graph already exists.")

        for i in range(number_of_vertices):
            self.add_vertex(i)
        for i in range(number_of_edges):
            source = randint(0, number_of_vertices - 1)
            destination = randint(0, number_of_vertices - 1)
            cost = randint(1, 100)
            self.add_edge(source, destination, cost)

    def read_graph_from_file(self, file_name):
        with open(file_name, 'r') as file:
            self.number_of_edges = 0
            self.number_of_vertices = 0
            self.inbound_edges = {}
            self.outbound_edges = {}
            self.costs = {}

            number_of_vertices, number_of_edges = file.readline().split()
            number_of_edges = int(number_of_edges)
            number_of_vertices = int(number_of_vertices)

            for vertex in range(number_of_vertices):
                self.add_vertex(vertex)

            while number_of_edges:
                line = file.readline().split()
                source, destination, cost = line[0], line[1], line[2]
                self.add_edge(int(source), int(destination), int(cost))
                number_of_edges -= 1

    def write_graph_to_file(self, file_name):
        with open(file_name, 'w') as file:
            file.write(f"{self.number_of_vertices} {self.number_of_edges}\n")
            for vertex, outbound_edges in self.outbound_edges.items():
                for outbound_edge in outbound_edges:
                    file.write(f"{vertex} {outbound_edge} {self.costs[(vertex, outbound_edge)]}\n")

File: First year/Second Semester/Graphs/A1/A1 - C++/src/directed_graph.cpp
#include "directed_graph.h"
#include "graph_exceptions.h"
using namespace std;

DirectedGraph::DirectedGraph(int number_of_vertices): number_of_vertices(number_of_vertices), number_of_edges(0){
    inbound_edges = {};
    outbound_edges = {};
    costs = {};
}

string DirectedGraph::toString() {
    string result = "DirectedGraph: " + to_string(number_of_vertices) +
                         " vertices, " + to_string(number_of_edges) + " edges\n";
    for (const auto& [vertex, _inbound_edges] : inbound_edges) {
        for (int inbound_edge : _inbound_edges) {
            if (this->inbound_edges[vertex].empty() && this->outbound_edges[vertex].empty()) {
                result += " " + to_string(vertex) + " -> isolated vertex\n";
                continue;
            }
            result += " " + to_string(inbound_edge) + " -> " + to_string(vertex) +
                      " ->: " + to_string(costs[{inbound_edge, vertex}]) + "\n";
        }
    }
    return result;
}

int DirectedGraph::GetNumberOfVertices() const{
    return number_of_vertices;
}

void DirectedGraph::SetNumberOfVertices(int numberOfVertices) {
    DirectedGraph::number_of_vertices = numberOfVertices;
}

bool DirectedGraph::CheckIfEdgeExists(int source, int destination) const {
    return costs.find({source, destination}) != costs.end();
}

int DirectedGraph::GetCostOfAnEdge(int source, int destination) const {
    if (CheckIfEdgeExists(source, destination))
        return costs.at({source, destination});
    else {
        throw GraphException("There is no edge from " + to_string(source) + " to " + to_string(destination));
    }
}

void DirectedGraph::SetCostOfAnEdge(int source, int destination, int cost) {
    // Check if the edge exists
    if (CheckIfEdgeExists(source, destination)) {
        // Update the cost if the edge exists
        costs[make_pair(source, destination)] = cost;
    } else {
        // Add the edge with the specified cost if it doesn't exist
        AddEdge(source, destination, cost);
    }
}

vector<int> DirectedGraph::GetInboundEdges(int vertex) const {
    // Check if the vertex exists in the graph
    auto it = inbound_edges.find(vertex);
    if (it == inbound_edges.end()) {
        throw GraphException("Vertex does not exist in the graph");
    }

    // Return the inbound edges for the vertex
    return it->second;
}

vector<int> DirectedGraph::GetOutboundEdges(int vertex) const {
    // Check if the vertex exists in the graph
    auto it = outbound_edges.find(vertex);
    if (it == outbound_edges.end()) {
        throw GraphException("Vertex does not exist in the graph");
    }

    // Return the outbound edges for the vertex
    return it->second;
}

vector<int> DirectedGraph::GetVertices() const {
    vector<int> vertices;
    vertices.reserve(inbound_edges.size());
    for (const auto& entry : inbound_edges) {
        vertices.push_back(entry.first);
    }
    return vertices;
}

bool DirectedGraph::CheckIfGraphIsEmpty() const {
    return number_of_vertices == 0;
}

void DirectedGraph::AddVertex(int vertex) {
    if (inbound_edges.find(vertex) != inbound_edges.end()) {
        throw GraphException("The vertex already exists");
    }

    inbound_edges[vertex] = {};
    outbound_edges[vertex] = {};
    ++number_of_vertices;
}

void DirectedGraph::RemoveVertex(int vertex) {
    if (inbound_edges.find(vertex) == inbound_edges.end()) {
        throw GraphException("The vertex does not exist");
    }

    for (const auto& [source, _outbound_edges] : outbound_edges) {
        if (CheckIfEdgeExists(source, vertex)) {
            RemoveEdge(source, vertex);
        }
    }

    for (const auto& [destination, _inbound_edges] : inbound_edges) {
        if (CheckIfEdgeExists(vertex, destination)) {
            RemoveEdge(vertex, destination);
        }
    }

    inbound_edges.erase(vertex);
    outbound_edges.erase(vertex);
    --number_of_vertices;
}

void DirectedGraph::AddEdge(int source, int destination, int cost) {
    if (CheckIfEdgeExists(source, destination)) {
        throw GraphException("The edge already exists");
    }

    try {
        AddVertex(source);
    } catch (const GraphException&) {
    }

    try {
        AddVertex(destination);
    } catch (const GraphException&) {
    }

    if (outbound_edges.find(source) == outbound_edges.end()) {
        outbound_edges[source] = {};
    }
    if (inbound_edges.find(destination) == inbound_edges.end()) {
        inbound_edges[destination] = {};
    }

    inbound_edges[destination].push_back(source);
    outbound_edges[source].push_back(destination);
    costs[{source, destination}] = cost;
    ++number_of_edges;
}

void DirectedGraph::RemoveEdge(int source, int destination) {
    if (!CheckIfEdgeExists(source, destination)) {
        throw GraphException("The edge does not exist");
    }

    inbound_edges[destination].erase(remove(inbound_edges[destination].begin(), inbound_edges[destination].end(), source),
                                     inbound_edges[destination].end());
    outbound_edges[source].erase(remove(outbound_edges[source].begin(), outbound_edges[source].end(), destination),
                                  outbound_edges[source].end());
    costs.erase({source, destination});
    --number_of_edges;
}

shared_ptr<DirectedGraph> DirectedGraph::GetCopyOfGraph() const {
    auto graph_copy = make_shared<DirectedGraph>(number_of_vertices);
    graph_copy->inbound_edges = inbound_edges;
    graph_copy->outbound_edges = outbound_edges;
    graph_copy->costs = costs;
    return graph_copy;
}

void DirectedGraph::CreateRandomGraph(int numberOfVertices, int numberOfEdges) {
    if (numberOfEdges > numberOfVertices * (numberOfVertices - 1)) {
        throw GraphException("Invalid input! The number of edges must be less than the number of vertices * (number of vertices - 1)");
    }
    if (this->number_of_vertices != 0) {
        throw GraphException("The graph already exists.");
    }

    random_device rd;
    mt19937 gen(rd());
    uniform_int_distribution<int> distribution(1, 100);

    for (int i = 0; i < numberOfVertices; ++i) {
        AddVertex(i);
    }

    uniform_int_distribution<int> cost_distribution(1, 100);
    // Generate all possible edges
    vector<pair<int, int>> possible_edges;
    for (int i = 0; i < numberOfVertices; ++i) {
        for (int j = 0; j < numberOfVertices; ++j) {
            if (i != j) {
                possible_edges.emplace_back(i, j);
            }
        }
    }

    int num_edges = numberOfEdges;
    for (const auto& edge : possible_edges) {
        int source = edge.first;
        int destination = edge.second;
        int cost = cost_distribution(gen); // Random cost between 1 and 100
        AddEdge(source, destination, cost);
        num_edges--;
        if (num_edges == 0) {
            break;
        }
    }
}

void DirectedGraph::ReadGraphFromFile(const string &file_name) {
    ifstream file(file_name);
    if (!file.is_open()) {
        throw GraphException("Unable to open file");
    }

    number_of_edges = 0;
    number_of_vertices = 0;
    inbound_edges.clear();
    outbound_edges.clear();
    costs.clear();

    int num_vertices, num_edges;
    file >> num_vertices >> num_edges;

    for (int vertex = 0; vertex < num_vertices; ++vertex) {
        AddVertex(vertex);
    }

    int source, destination, cost;
    while (num_edges--) {
        file >> source >> destination >> cost;
        AddEdge(source, destination, cost);
    }
}

void DirectedGraph::WriteGraphToFile(const string &file_name) const {
    ofstream file(file_name);
    if (!file.is_open()) {
        throw GraphException("Unable to open file for writing");
    }

    file << number_of_vertices << " " << number_of_edges << "\n";

    for (const auto& [vertex, outboundEdges] : outbound_edges) {
        for (int outbound_edge : outboundEdges) {
            file << vertex << " " << outbound_edge << " " << costs.at({vertex, outbound_edge}) << "\n";
        }
    }
}




